flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java

```diff
@@ -0,0 +1,120 @@
+  private final List<String> taskNames;
+
+  private transient Map<Integer, Counter> successfulStreamResultCounterMap;
+  private transient Map<Integer, Counter> failedStreamResultCounterMap;
+  private transient Map<Integer, AtomicLong> lastRunLength;
+  private transient TriggerLockFactory.Lock lock;
+  private transient TriggerLockFactory.Lock recoveryLock;
+  private transient long lastProcessed = 0L;
+
+  public LockRemover(TriggerLockFactory lockFactory, List<String> taskNames) {
+    Preconditions.checkNotNull(lockFactory, "Lock factory should no be null");
+    Preconditions.checkArgument(
+        taskNames != null && !taskNames.isEmpty(), "Invalid task names: null or empty");
+
+    this.lockFactory = lockFactory;
+    this.taskNames = taskNames;
+  }
+
+  @Override
+  public void open() throws Exception {
+    super.open();
+    this.successfulStreamResultCounterMap = Maps.newHashMapWithExpectedSize(taskNames.size());
+    this.failedStreamResultCounterMap = Maps.newHashMapWithExpectedSize(taskNames.size());
+    this.lastRunLength = Maps.newHashMapWithExpectedSize(taskNames.size());
+    for (int i = 0; i < taskNames.size(); ++i) {
+      String name = taskNames.get(i);
+      successfulStreamResultCounterMap.put(
+          i,
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.SUCCESSFUL_STREAM_COUNTER));
+      failedStreamResultCounterMap.put(
+          i,
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.FAILED_STREAM_COUNTER));
+      AtomicLong length = new AtomicLong(0);
+      lastRunLength.put(i, length);
+      getRuntimeContext()
+          .getMetricGroup()
+          .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+          .gauge(TableMaintenanceMetrics.LAST_RUN_LENGTH, length::get);
+    }
+
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+  }
+
+  @Override
+  public void processElement(StreamRecord<TaskResult> element) {
+    TaskResult value = element.getValue();
+    LOG.debug("TaskResult {} arrived", value);
+    long length = System.currentTimeMillis() - value.startEpoch();
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                value.startEpoch(), value.taskId(), length, value.success(), value.exceptions())));
+    lock.unlock();
+    this.lastProcessed = value.startEpoch();
+
+    // Update the metrics
+    lastRunLength.get(value.taskId()).set(length);
+    if (value.success()) {
+      successfulStreamResultCounterMap.get(value.taskId()).inc();
+    } else {
+      failedStreamResultCounterMap.get(value.taskId()).inc();
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) {
+    if (mark.getTimestamp() > lastProcessed) {
```

Comment:
should  this be `>` or `>=`?

Reply:
The assumption is:
- Every `TaskResult` is followed by a `Watermark` for normal `Trigger`s
- For the recovery `Trigger` there is no `processElement`

There are 3 possibilities:
- Normal execution - we receive a `TaskResult` and then a `Watermark` - unlocking the `lock` is handled by the `processElement`
- Recovery without ongoing execution (unlocking the `recoveryLock`) - we receive a `Watermark` without any `TaskResults` - unlocking the `recoveryLock` and a possible `lock` is handled by the `processWatermark` (the lastProcessed is `0`)
- Recovery with an ongoing execution - we receive a `TaskResult` and then a `Watermark` - unlocking the `lock` is handled by the `processElement`, unlocking the `recoveryLock` is handled by the `processWatermark` (the lastProcessed is the start time of the old task)

Reply:
can you add the above context as code comment here and in the processElement method?

Reply:
Added the comment to the class